pkg/driver/controller.go

```diff
@@ -0,0 +1,337 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package driver
+
+import (
+	"context"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/google/uuid"
+	"github.com/kubernetes-sigs/aws-efs-csi-driver/pkg/cloud"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/klog"
+)
+
+const (
+	AccessPointMode     = "efs-ap"
+	FsId                = "fileSystemId"
+	GidMin              = "gidRangeStart"
+	GidMax              = "gidRangeEnd"
+	DirectoryPerms      = "directoryPerms"
+	BasePath            = "basePath"
+	ProvisioningMode    = "provisioningMode"
+	DefaultGidMin       = 50000
+	DefaultGidMax       = 7000000
+	RootDirPrefix       = "efs-csi-ap"
+	TempMountPathPrefix = "/var/lib/csi/pv"
+)
+
+var (
+	// controllerCaps represents the capability of controller service
+	controllerCaps = []csi.ControllerServiceCapability_RPC_Type{
+		csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME,
+	}
+)
+
+func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+	klog.V(4).Infof("CreateVolume: called with args %+v", *req)
+	volName := req.GetName()
+	if volName == "" {
+		return nil, status.Error(codes.InvalidArgument, "Volume name not provided")
+	}
+
+	// Volume size is required to match PV to PVC by k8s.
+	// Volume size is not consumed by EFS for any purposes.
+	volSize := req.GetCapacityRange().GetRequiredBytes()
+
+	volCaps := req.GetVolumeCapabilities()
+	if len(volCaps) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume capabilities not provided")
+	}
+
+	if !d.isValidVolumeCapabilities(volCaps) {
+		return nil, status.Error(codes.InvalidArgument, "Volume capabilities not supported")
+	}
+
+	var (
+		gidMin           int
+		gidMax           int
+		basePath         string
+		provisioningMode string
+		err              error
+	)
+
+	//Parse parameters
+	volumeParams := req.GetParameters()
+	if value, ok := volumeParams[ProvisioningMode]; ok {
+		provisioningMode = value
+		//TODO: Add FS provisioning mode check when implemented
+		if provisioningMode != AccessPointMode {
+			errStr := "Provisioning mode " + provisioningMode + " is not supported. Only Access point provisioning: 'efs-ap' is supported"
+			return nil, status.Error(codes.InvalidArgument, errStr)
+		}
+	} else {
+		return nil, status.Errorf(codes.InvalidArgument, "Missing %v parameter", ProvisioningMode)
```

Comment:
would it be better to default it to efs-ap?

Reply:
For now defaulting would be okay, however, we would have to bring this check back after we introduce 'efs-fs'. So lets keep it as is.

Reply:
okay, as long as we provide an example storageclass and document that it is required then I dont mind